src/app/crud/base.py

```diff
@@ -0,0 +1,89 @@
+from typing import Any, Dict, Generic, List, Optional, Type, TypeVar, Union
+
+from fastapi.encoders import jsonable_encoder
+from pydantic import BaseModel
+from sqlalchemy.orm import Session
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from app.db.base_class import Base
+from sqlalchemy.sql.expression import select
+
+ModelType = TypeVar("ModelType", bound=Base)
+CreateSchemaType = TypeVar("CreateSchemaType", bound=BaseModel)
+UpdateSchemaType = TypeVar("UpdateSchemaType", bound=BaseModel)
+
+
+class CRUDBase(Generic[ModelType, CreateSchemaType, UpdateSchemaType]):
+    def __init__(self, model: Type[ModelType]):
+        """
+        CRUD object with default methods to Create, Read, Update, Delete (CRUD).
+        **Parameters**
+        * `model`: A SQLAlchemy model class
+        * `schema`: A Pydantic model (schema) class
+        """
+        self.model = model
+
+
+    async def get(self, db: AsyncSession, id: Any) -> Optional[ModelType]:
+        result = await db.execute(select(self.model).filter(self.model.id == id))
+        return result.scalars().first()
+
+
+    async def get_multi(
+        self, db: AsyncSession, *, skip: int = 0, limit: int = 100
+    ) -> List[ModelType]:
+        query = select(self.model).offset(skip).limit(limit)
+        result = await db.execute(query)
+        res = result.scalars().all()
+        return res
+
+
+    async def get_multi_by_owner(
+        self, db: AsyncSession, *, owner_id: int, skip: int = 0, limit: int = 100
+    ) -> List[ModelType]:
+        result =  await db.execute(
+            select(self.model)
+            .filter(self.model.owner_id == owner_id)
+            .offset(skip)
+            .limit(limit)
+        )
+        return result.scalars().all()
+
+
+    async def create(self, db: AsyncSession, *, obj_in: CreateSchemaType) -> ModelType:
+        #obj_in_data = jsonable_encoder(obj_in)
+        obj_in_data = obj_in.dict()
+        db_obj = self.model(**obj_in_data)  # type: ignore
+        db.add(db_obj)
+        await db.commit()
+        await db.refresh(db_obj)
+        return db_obj
+
+
+    async def update(
+        self,
+        db: AsyncSession,
+        *,
+        db_obj: ModelType,
+        obj_in: Union[UpdateSchemaType, Dict[str, Any]]
+    ) -> ModelType:
+        obj_data = jsonable_encoder(db_obj)
+        if isinstance(obj_in, dict):
+            update_data = obj_in
+        else:
+            update_data = obj_in.dict(exclude_unset=True)
+        for field in obj_data:
+            if field in update_data:
+                setattr(db_obj, field, update_data[field])
+        db.add(db_obj)
+        await db.commit()
+        await db.refresh(db_obj)
+        return db_obj
+
+
+    async def remove(self, db: AsyncSession, *, id: int) -> ModelType:
```

Comment:
I suggest we don't call `remove` with and ID but with `ModelType` (which the caller must have retrieved from the database).
Also considering the workflow we have in `robot_type.py`we end up calling `get` twice instead of just once.

Reply:
Nice catch! Changed it to ModelType.